rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp

```diff
@@ -75,7 +75,7 @@ class MultiThreadedExecutor : public rclcpp::Executor
 
 protected:
   RCLCPP_PUBLIC
-  void
+  virtual void
```

Comment:
> The run method should be override'able by a child class.

Can we just not create `UserMultiThreadedExecutor` inherited from `MultiThreadedExecutor` to override `run` method and then call `UserMultiThreadedExecutor::spin()`?

i think adding `virtual` here is supposed to mean that, having sub-class instance pointed by super-class pointer, and handle super-class pointers in the program which actually calls sub-class `run` override method as super-class? i do not really think of the use case for this.

i might be mistaken...

Reply:
No we cannot really do that because since the run method is not virtual, the `MultiThreadedExecutor::spin()` will just call the `MultiThreadedExecutor::run(...)` method (as opposed to the child class implementation). 

Making this method virtual allows for the child class to be executed correctly. 
Here is an example of a child class implementation:
```
class SafeMultiThreadedExecutor : public  rclcpp::executors::MultiThreadedExecutor                                                                                                                                 
{                                                                                                                                                                                                                  
public:                                                                                                                                                                                                            
  SafeMultiThreadedExecutor() : rclcpp::executors::MultiThreadedExecutor()                                                                                                                                         
  {}                                                                                                                                                                                                                                                                                                                                                                                                                               
protected:                                                                                                                                                                                                         
  void run(size_t this_thread_number) override                                                                                                                                                                     
  {                                                                                                                                                                                                                
    try {                                                                                                                                                                                                          
      rclcpp::executors::MultiThreadedExecutor::run(this_thread_number);                                                                                                                                           
    } catch (const std::exception & e) {                                                                                                                                                                           
      RCUTILS_LOG_ERROR_NAMED(                                                                                                                                                                                     
        "api_relay", "Exception in MultiThreadedExecutor (%ld): %s", this_thread_number, e.what());                                                                                                                
    }                                                                                                                                                                                                              
  }                                                                                                                                                                                                                
};             
```

Note that the `override` keyword tells the compiler that this implementation is expected to override the base class implementation. Without my patch here, the compiler will fail indicating that this class does not override the base class implementation. 

Take a look at the [spin()](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/executors/multi_threaded_executor.hpp#L68-L70) method. This is how it should be implemented...

Reply:
I think that your example should rather be implemented by overriding `spin` and wrapping the `run` calls with the try-catch
https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L69

Reply:
That is another approach but it does force me to replicate the entire spin method. The run invocation that is problematic in my case is [this one](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L64) That's because those calls are run in a separate thread context. So if an exception is thrown in those other thread, the application will terminate abnormally with no possibility to handle the exception. 

Obviously, the whole spin method can be copied but I was hoping for a more lightweight approach...

Reply:
Ah, i see what you mean. instead of `virtual`, it cannot call sub-class `run` in `MultiThreadedExecutor::spin()`. for your example, even calling `SafeMultiThreadedExecutor::spin()` (which is calling `MultiThreadedExecutor::spin()`) that will call `MultiThreadedExecutor::run()`.

> That is another approach but it does force me to replicate the entire spin method.

true.